packages/core/src/context.ts

```diff
@@ -0,0 +1,54 @@
+import { RuntimeEnvironment } from './type';
+import { TasksLoader } from './task-loader';
+import { ExtenderManager } from './extenders';
+import { KuaiError } from './errors';
+import { ERRORS } from './errors-list';
+
+export class KuaiContext {
```

Comment:
Maybe only `createKuaiContext` need to be set to `static`.

Reply:
If `getKuaiContext` is not static, how should I call it, or should I remove`getKuaiContext` and use `__KuaiContext` directly?

Reply:
Is it possible to set and get the singleton in one action like

```typescript
getKuaiContext() {
  if (!this.__KuaiContext) {
    this.__KuaiContext = new KuaiContext()
  }
  return this.__KuaiContext
}
```

Reply:
After create, some initialization work will be done (such as setting runtime) If use `getOrCreate`, there may be a problem that kuaiContext is created without initialization, which may generate some errors

Reply:
I didn't get the point here. The code referenced is

```typescript
export function initialKuai(args: KuaiArguments = {}): KuaiContext {
  loadTsNode();

  const ctx = KuaiContext.createKuaiContext();

  const { config } = loadConfigAndTasks(args);

  const env = new KuaiRuntimeEnvironment(config, ctx.tasksLoader.getTasks(), ctx.extendersManager.getExtenders());
  ctx.setRuntimeEnvironment(env);

  return ctx;
}
```
After expansion it would be 
```typescript
export function initialKuai(args: KuaiArguments = {}): KuaiContext {
  loadTsNode();

  const ctx = (() => {
    if (this.__KuaiContext !== undefined) {
      throw new KuaiError(ERRORS.GENERAL.CONTEXT_ALREADY_CREATED);
    }
    const _ctx = new KuaiContext();
    this.__KuaiContext = _ctx;
    return _ctx;
  })()

  const { config } = loadConfigAndTasks(args);

  const env = new KuaiRuntimeEnvironment(config, ctx.tasksLoader.getTasks(), ctx.extendersManager.getExtenders());
  ctx.setRuntimeEnvironment(env);

  return ctx;
}
```

which is exactly the same as 
```typescript
export function initialKuai(args: KuaiArguments = {}): KuaiContext {
  loadTsNode();

  const ctx = new KuaiContext()

  const { config } = loadConfigAndTasks(args);

  const env = new KuaiRuntimeEnvironment(config, ctx.tasksLoader.getTasks(), ctx.extendersManager.getExtenders());
  ctx.setRuntimeEnvironment(env);

  return ctx;
}
```

If the get and set are done in one action, the code would be expanded into the same logic here, unless the `context` has been initialized before and an exception `new KuaiError(ERRORS.GENERAL.CONTEXT_ALREADY_CREATED)` will be thrown

Reply:
ok, got

Reply:
> In some case where just want to get the `KuaiContext` and don't know the `KuaiArguments` (usually from the cli argv / env / config file)
> 
> For example
> 
> https://github.com/ckb-js/kuai/blob/d34bdf93f7b9a1521254549cc69c0fab6f67a117/packages/core/src/config/config-env.ts#L37-L43

It's expected to throw an exception `new KuaiError(ERRORS.GENERAL.CONTEXT_NOT_CREATED)` if `initialKuai` is not done, am I right?

If yes, then I got it.

Reply:
But it's still confusing because the logic here is coupled. Accessibility of KuaiContext instance relies on other functions.

Reply:
yes

Reply:
How about only letting `getKuaiContext` and `createKuaiContext` static?
If after `createKuaiContext` should init environment, why not transfer as a parameter to `createKuaiContext`?
```typescript
  public static createKuaiContext(env: RuntimeEnvironment): KuaiContext {
    if (this.__KuaiContext !== undefined) {
      throw new KuaiError(ERRORS.GENERAL.CONTEXT_ALREADY_CREATED);
    }
    const ctx = new KuaiContext();
    ctx.setRuntimeEnvironment(env)
    this.__KuaiContext = ctx;
    return ctx;
  }
```

Reply:
There is nothing wrong with using `getOrCreate` here, except that if separate get and create, it is easier to find the error of not creating when using get, while using create makes the programmer aware of the need for some initialization processing here

Reply:
> There is nothing wrong with using `getOrCreate` here, except that if separate get and create, it is easier to find the error of not creating when using get, while using create makes the programmer aware of the need for some initialization processing here

The problem comes here.

The `getKuaiContext` is coupled with `initializeKuai` so if it's called alone, an exception `new KuaiError(ERRORS.GENERAL.CONTEXT_NOT_CREATED)` will be thrown.

However, I, as the user of this function, don't know `initializeKuai`(the message also doesn't cover it) and will only try `createKuaiContext` before `getKuaiContext`. What's worse is that `createKuaiContext` makes `getKuaiContext` work.

If it's fine to make `getKuaiContext` work by `createKuaiContext`, they would be good to be placed together.

If `createKuaiContext` is not the only condition to allow user to get the instance of `KuaiContext`, the check should be placed in the `get` method.

Reply:
got it